extensions/data-plane/data-plane-http/src/main/java/org/eclipse/edc/connector/dataplane/http/pipeline/HttpDataSink.java

```diff
@@ -37,23 +38,36 @@ public class HttpDataSink extends ParallelSink {
     @Override
     protected StatusResult<Void> transferParts(List<DataSource.Part> parts) {
         for (DataSource.Part part : parts) {
-            var request = params.toRequest(part::openStream);
-            try (var response = httpClient.execute(request)) {
-                if (!response.isSuccessful()) {
-                    monitor.severe(format("Error {%s: %s} received writing HTTP data %s to endpoint %s for request: %s",
-                            response.code(), response.message(), part.name(), request.url().url(), request));
-                    return ERROR_WRITING_DATA;
+            try (var is = part.openStream()) {
+                var result = transferPart(part.name(), is.readAllBytes());
+                if (result.failed()) {
+                    return result;
                 }
-
-                return StatusResult.success();
-            } catch (Exception e) {
-                monitor.severe(format("Error writing HTTP data %s to endpoint %s for request: %s", part.name(), request.url().url(), request), e);
+            } catch (IOException e) {
+                monitor.severe(format("Failed to open stream for part: %s", part.name()), e);
                 return ERROR_WRITING_DATA;
             }
         }
         return StatusResult.success();
     }
 
+    private StatusResult<Void> transferPart(String name, byte[] bytes) {
+        var request = params.toRequest(bytes);
+        try (var response = httpClient.execute(request)) {
+            if (!response.isSuccessful()) {
+                monitor.severe(format("Error {%s: %s} received writing HTTP data %s to endpoint %s for request: %s",
+                        response.code(), response.message(), name, request.url().url(), request));
+                return ERROR_WRITING_DATA;
```

Comment:
may want to attach the `response`'s message/error to the `ERROR_WRITING_DATA` result

Reply:
Don't know... Do we want to expose the response from the backend REST API targeted by the source?